File: app/src/main/java/me/jfenn/alarmio/data/preference/RingtonePreferenceData.java
package me.jfenn.alarmio.data.preference;

import me.jfenn.alarmio.R;
import me.jfenn.alarmio.data.PreferenceData;
import me.jfenn.alarmio.data.SoundData;
import me.jfenn.alarmio.dialogs.SoundChooserDialog;

public class RingtonePreferenceData extends CustomPreferenceData {

    private PreferenceData preference;

    public RingtonePreferenceData(PreferenceData preference, int name) {
        super(name);
        this.preference = preference;
    }

    @Override
    public String getValueName(ViewHolder holder) {
        String sound = preference.getValue(holder.getContext(), "");
        return sound != null && sound.length() > 0 ? SoundData.fromString(sound).getName() : holder.getContext().getString(R.string.title_sound_none);
    }

    @Override
    public void onClick(final ViewHolder holder) {
        SoundChooserDialog dialog = new SoundChooserDialog();
        dialog.setListener(sound -> {
            preference.setValue(holder.getContext(), sound != null ? sound.toString() : null);
            bindViewHolder(holder);
        });
        dialog.show(holder.getAlarmio().getFragmentManager(), null);
    }
}

File: app/src/main/java/me/jfenn/alarmio/data/preference/RingtonePreferenceData.kt
package me.jfenn.alarmio.data.preference

import me.jfenn.alarmio.R
import me.jfenn.alarmio.data.PreferenceData
import me.jfenn.alarmio.data.SoundData
import me.jfenn.alarmio.dialogs.SoundChooserDialog

/**
 * Allows the user to select from a set of
 * ringtone sounds (preference is a string
 * that can be recreated into a SoundData
 * object).
 */
class RingtonePreferenceData(private val preference: PreferenceData, name: Int) : CustomPreferenceData(name) {

    override fun getValueName(holder: CustomPreferenceData.ViewHolder): String {
        return preference.getValue(holder.context, "")?.let{ sound ->
            if (!sound.isEmpty())
                SoundData.fromString(sound)?.name ?: holder.context.getString(R.string.title_sound_none)
            else null
        } ?: holder.context.getString(R.string.title_sound_none)
    }

    override fun onClick(holder: CustomPreferenceData.ViewHolder) {
        holder.alarmio?.fragmentManager?.let { manager ->
            val dialog = SoundChooserDialog()
            dialog.setListener { sound ->
                preference.setValue(holder.context, sound?.toString())
                bindViewHolder(holder)
            }
            dialog.show(manager, null)
        }
    }
}
